metrics_prometheus/metrics.go

```diff
@@ -0,0 +1,193 @@
+package metrics_prometheus
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	mr "code.cloudfoundry.org/go-metric-registry"
+
+	"code.cloudfoundry.org/gorouter/config"
+	"code.cloudfoundry.org/gorouter/metrics"
+	"code.cloudfoundry.org/gorouter/route"
+)
+
+// Metrics represents a prometheus metrics endpoint.
+type Metrics struct {
+	RouteRegistration           mr.CounterVec
+	RouteUnregistration         mr.CounterVec
+	RoutesPruned                mr.Counter
+	TotalRoutes                 mr.Gauge
+	TimeSinceLastRegistryUpdate mr.Gauge
+	RouteLookupTime             mr.Histogram
+	RouteRegistrationLatency    mr.Histogram
+	RoutingRequest              mr.CounterVec
+	BadRequest                  mr.Counter
+	BadGateway                  mr.Counter
+	EmptyContentLengthHeader    mr.Counter
+	BackendInvalidID            mr.Counter
+	BackendInvalidTLSCert       mr.Counter
+	BackendTLSHandshakeFailed   mr.Counter
+	BackendExhaustedConns       mr.Counter
+	WebsocketUpgrades           mr.Counter
+	WebsocketFailures           mr.Counter
+	Responses                   mr.CounterVec
+	RouteServicesResponses      mr.CounterVec
+	RoutingResponseLatency      mr.HistogramVec
+	perRequestMetricsReporting  bool
+}
+
+func NewMetricsRegistry(config config.PrometheusConfig) *mr.Registry {
+	var metricsRegistry *mr.Registry
+	if config.Port != 0 && config.CertPath != "" {
```

Comment:
This allows `config.Port == 0` when cert path is not set, which would start on a random port?

Either way the distinction for TLS or not should be on CertPath, not on port, I think.

Reply:
We are using port = 0 in our tests , in order to start the test on the random port. Otherwise it is not possible to run the tests.
In the routing release we added a check which does not allow to set port = 0.